// LANGUAGE: +ValhallaValueClasses
// IGNORE_BACKEND_K1: ANY
// RUN_PIPELINE_TILL: FRONTEND
// JVM_TARGET: 23
// EXPLICIT_API_MODE: STRICT

public <!VALUE_CLASS_WITHOUT_JVM_INLINE_ANNOTATION!>value<!> class FooValue(<!NO_EXPLICIT_VISIBILITY_IN_API_MODE!>val i<!>: Int)
<!NO_EXPLICIT_VISIBILITY_IN_API_MODE!><!VALUE_CLASS_WITHOUT_JVM_INLINE_ANNOTATION!>value<!> class FooValue2<!>(<!NO_EXPLICIT_VISIBILITY_IN_API_MODE!>val i<!>: Int)
