// LANGUAGE: +ValhallaValueClasses
// IGNORE_BACKEND_K1: ANY
// RUN_PIPELINE_TILL: FRONTEND
// JVM_TARGET: 23
// ENABLE_JVM_PREVIEW

<!VALUE_CLASS_WITHOUT_JVM_INLINE_ANNOTATION!>value<!> class A<!INLINE_CLASS_CONSTRUCTOR_WRONG_PARAMETERS_SIZE!>(val y: Int, val z: A)<!>
<!VALUE_CLASS_WITHOUT_JVM_INLINE_ANNOTATION!>value<!> class B<!INLINE_CLASS_CONSTRUCTOR_WRONG_PARAMETERS_SIZE!>(val y: Int, val z: B?)<!>

<!VALUE_CLASS_WITHOUT_JVM_INLINE_ANNOTATION!>value<!> class C<!INLINE_CLASS_CONSTRUCTOR_WRONG_PARAMETERS_SIZE!>(val a: C?, val b: C, val c: C, val d: C?, val e: D)<!>

<!VALUE_CLASS_WITHOUT_JVM_INLINE_ANNOTATION!>value<!> class D(val a: <!VALUE_CLASS_CANNOT_BE_RECURSIVE!>C<!>)

<!INLINE_CLASS_DEPRECATED!>inline<!> class Inline1(val x: <!VALUE_CLASS_CANNOT_BE_RECURSIVE!>Inline1<!>)
<!INLINE_CLASS_DEPRECATED!>inline<!> class Inline2(val x: <!VALUE_CLASS_CANNOT_BE_RECURSIVE!>Inline2?<!>)
<!INLINE_CLASS_DEPRECATED!>inline<!> class Inline3(val x: <!VALUE_CLASS_CANNOT_BE_RECURSIVE!>Inline4<!>)
<!INLINE_CLASS_DEPRECATED!>inline<!> class Inline4(val x: <!VALUE_CLASS_CANNOT_BE_RECURSIVE!>Inline3?<!>)
